### DISCUSSION TITLE
Support alternative databases next to Postgres, e.g. MySQL

### DISCUSSION THREAD

**Discussion by zoltan-fedor** [langfuse/langfuse#976] [Ideas]
As the title says, it would be nice if there would be other options for backend database outside of just Postgres.
Once such option could be MySQL.

**Comment by marcklingen** [comment]
Hello @zoltan-fedor, could you share why you need support for additional databases like MySQL? I'd imagine this would be beneficial if you're running an existing MySQL db that you'd like to repurpose for your Langfuse instance. However, for production deployments, an independent database for Langfuse is always recommended.

Adding support for extra databases is most likely not be the best use of resources for now. Primarily because the Postgres SQL dialect is used across the application and translating between dialects adds a new source of potential errors/regressions.

**Comment by zoltan-fedor** [comment]
Hi @marcklingen,
You are correct, the reason why it would be beneficial, so those like us who are already running a MySQL database wouldn't be required to also run a Postgres one too.

Multi-database backend support is why most applications are written without hard-coding SQL dialects, but rather writing with ORM tools like SQLAlchemy - which allows for the use of multiple different backend databases.
You are right, if Langfuse is written with hardcoded SQL without an ORM, then that SQL might be made specific to Postgres, which I suspect is going to be a problem in the long term. All successful open source projects I have ever participated in has eventually switched to using an ORM, so it can support multiple backend databases.

I also hear you about the dedicated database recommendation - which is again an odd design choice for the long term. I can understand the appeal of it - so you can give root grants to the app to create new databases, but in generally not a good long-term choice as you (eventually) will have users who want to use multi-tenant databases.

I am sorry to hear that support for "extra databases" is not considered right now.

**Comment by marcklingen** [comment]
> You are correct, the reason why it would be beneficial, so those like us who are already running a MySQL database wouldn't be required to also run a Postgres one too.

Thanks for the details, super helful.

> Multi-database backend support is why most applications are written without hard-coding SQL dialects, but rather writing with ORM tools like SQLAlchemy - which allows for the use of multiple different backend databases.
You are right, if Langfuse is written with hardcoded SQL without an ORM, then that SQL might be made specific to Postgres, which I suspect is going to be a problem in the long term. All successful open source projects I have ever participated in has eventually switched to using an ORM, so it can support multiple backend databases.

Langfuse uses [Prisma](https://github.com/prisma/prisma) as an ORM but for performance reasons SQL is used directly in some places – for example in the ingestion API to reduce concurrency issues via Postgres transactions/upserts and for the UI and GET API as Postgres views are used for standardized metrics which are derived from the ingested data.

> I also hear you about the dedicated database recommendation - which is again an odd design choice for the long term. I can understand the appeal of it - so you can give root grants to the app to create new databases, but in generally not a good long-term choice as you (eventually) will have users who want to use multi-tenant databases.

The application at runtime does not use root access to the database as the db is only migrated when upgrading the container.

Multi-tenancy within Postgres should be possible by defining a schema as part of the Postgres connection string.

> I am sorry to hear that support for "extra databases" is not considered right now.

I am open to discuss this if there is demand by more teams and someone is willing to contribute to this change. At the moment it does not create much value for existing users which is why I cannot directly commit some of my own time, always open to reconsider though! Hope you understand

**Reply by zoltan-fedor** [reply]
Thanks for being open to discuss it and potentially reconsider it in the future.

Let's see if more teams will be interested in it.
Let's leave this discussion open to see if there will be additional interest in this in the future.

**Reply by marcklingen** [reply]
Sounds good, hope this is not a big blocker for you and you find a way around it

**Reply by zoltan-fedor** [reply]
Thanks. No worries, we can always just run Postgres as a dedicated database for Langfuse, it is just we are trying to minimize the number of databases we need to run.

**Comment by rseymour** [comment]
The best reason is that Postgres has one missing character in its ASCII/UTF-8 support for strings, so if a string has the control character NULL `\0` `\u0000` in it the Json or string (text types, which includes jsonb) will never be inserted with a normal insert statement.  I haven't come across many responses from LLMs including it, but sometimes requests do if they aren't filtered.  

More details here: https://www.postgresql.org/docs/current/datatype-json.html I generally use Postgres in security contexts where NULL in strings (on purpose) is a common occurrence.

**Comment by serlina** [comment]
I think our team also has the requirement to support multiple databases, because we already use mongodb as our database. for the new involved system "langfuse", we have to maintain a new database postgres,  so it would be much better to support extral databases, thanks

**Reply by marcklingen** [reply]
Unfortunately, support for mongo is kind of impossible as it is a document db that is not well suited to model the data stored in Langfuse. You'll need to setup a postgres db or set up a free account on Langfuse Cloud: cloud.langfuse.com

**Comment by zaji-xuan** [comment]
I also agree with the author's idea, because if PostgreSQL's disk is full, cleaning it up will be very troublesome

**Reply by marcklingen** [reply]
Thanks for the +1, what is the trouble of removing old traces from postgres? more details on how to do this here: https://github.com/orgs/langfuse/discussions/2268#discussioncomment-10201896

**Comment by jlevine-intermedia** [comment]
For organizations with big existing IT footprints, adding another database type means adding another set of DBAs (or retraining the DBAs we already have). As an org with lots of MariaDB/MySQL already, we'd rather spin up yet another of the same database than have a maze of twisty little databases, all different (with apologies to Will Crowther)... so for me, +1 vote for database independence. 

> I am open to discuss this if there is demand by more teams and someone is willing to contribute to this change. 

Depending on how much effort we're talking about, we might be interested in this.
